plutarch-extra/Plutarch/Extra/Api.hs

```diff
@@ -0,0 +1,90 @@
+module Plutarch.Extra.Api (
+  getContinuingOutputs,
+  findOwnInput,
+  findDatum,
+) where
+
+import GHC.Records (HasField (getField))
+
+import Plutarch.Api.V1 (
+  PAddress,
+  PDatum,
+  PDatumHash,
+  PScriptContext,
+  PScriptPurpose (PSpending),
+  PTuple,
+  PTxInInfo,
+  PTxInfo,
+  PTxOut,
+  PTxOutRef,
+ )
+import Plutarch.Prelude
+
+import Plutarch.Extra.Monad (tmatch)
+
+-- | Find the output txns corresponding to the input being validated.
+getContinuingOutputs :: Term s (PScriptContext :--> PBuiltinList PTxOut)
+getContinuingOutputs = phoistAcyclic $
+  plam $ \sc -> unTermCont $ do
+    let txinfo = pfield @"txInfo" # sc
+    tmatch (findOwnInput # sc) >>= \case
+      PJust te -> do
+        let outs = pfield @"outputs" # txinfo
+            resolved = pfield @"resolved" # te
+            outAdr = pfield @"address" # resolved
+        pure $ pfilter # (matches # outAdr) #$ pmap # plam pfromData # outs
+      PNothing ->
+        pure $ ptraceError "can't get any continuing outputs"
+  where
+    matches :: Term s (PAddress :--> PTxOut :--> PBool)
+    matches = phoistAcyclic $
+      plam $ \adr txOut ->
+        adr #== pfield @"address" # txOut
+
+{- | Find the input currently being validated.
+
+  Tries to finds the transaction's input by looking for a `PTxInInfo` in the inputs
+  coresponding to the `PTxOutRef` which the script purpose is spending
+-}
+findOwnInput :: Term s (PScriptContext :--> PMaybe PTxInInfo)
+findOwnInput = phoistAcyclic $
+  plam $ \sc -> unTermCont $ do
+    ctx <- tcont $ pletFields @["txInfo", "purpose"] sc
+    tmatch (getField @"purpose" ctx) >>= \case
+      PSpending outRef' -> do
+        let outRef = pfield @"_0" # outRef'
+        pure $
+          pfind # (matches # outRef)
+            #$ pmap # plam pfromData
+            #$ pfromData
+            $ pfield @"inputs" # (getField @"txInfo" ctx)
+      _ ->
+        pure $ pcon PNothing
+  where
+    matches :: Term s (PTxOutRef :--> PTxInInfo :--> PBool)
+    matches = phoistAcyclic $
+      plam $ \outref txininfo ->
+        pfield @"id" # outref
+          #== pfield @"id" # (pfield @"outRef" # txininfo)
+
+-- | Find the data corresponding to a data hash, if there is one
+findDatum :: Term s (PDatumHash :--> PTxInfo :--> PMaybe PDatum)
```

Comment:
If going for the "maximally focused input" route, this should take in the datum assoc list instead of the whole txInfo